### PR TITLE
SingleRun and MultipleRun Deterministic

### PR DESCRIPTION
**Pull request by andychlee**
SingleRunDeterministic and MultipleRunDeterministic is not correctly checking for calls to methods.
I think I've included all the necessary logic for the BaseTypeVisitor to check, but it doesn't.

### CODE REVIEW

File: .travis-build.sh [hunk 1]
```diff
@@ -35,21 +35,22 @@ export SHELLOPTS
 # command to rebuild it again.
 
 if [[ "${GROUP}" == "junit" || "${GROUP}" == "all" ]]; then
-  (cd checker && ant junit-tests-nojtreg-nobuild)
+  (cd framework && ant -e -find build.xml flow-tests)
```
Comment (wmdietl) [inline .travis-build.sh]: Don't forget to undo these changes to the build before the final PR.

File: .travis.yml [hunk 2]
```diff
@@ -13,9 +13,6 @@ services:
 # jdkany contains extra packages that permit it to be used to build the manual
 env:
   matrix:
-  - JDKVER=jdk7 GROUP=junit
```
Comment (wmdietl) [inline .travis.yml]: Also this change should be undone.

File: checker/examples/units-extension/qual/FrequencyRelations.java [hunk 3]
```diff
@@ -34,7 +34,8 @@ public UnitsRelations init(ProcessingEnvironment env) {
     }
 
     /** No multiplications yield Hertz. */
-    public /*@Nullable*/ AnnotationMirror multiplication(AnnotatedTypeMirror lht, AnnotatedTypeMirror rht) {
+    public /*@Nullable*/ AnnotationMirror multiplication(
```
Comment (wmdietl) [inline checker/examples/units-extension/qual/FrequencyRelations.java]: There seem to be many changes to formatting. These shouldn't show up, as both sides should already have them. Can you see how to get rid of these differences?

File: dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java [hunk 4]
```diff
@@ -95,12 +96,16 @@ public static PurityResult checkPurity(
         protected final List<Pair<Tree, String>> notSeFreeReasons;
         protected final List<Pair<Tree, String>> notDetReasons;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java]: I assume that you'll remove this field when you remove the corresponding methods.

File: dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java [hunk 5]
```diff
@@ -357,6 +388,10 @@ public PurityResult visitMethodInvocation(MethodInvocationTree node, PurityResul
                     p.addNotDetReason(node, reason);
                 } else if (!seFree) {
                     p.addNotSeFreeReason(node, reason);
+                } else if (!sdet) {
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java]: Are you sure that the sdet and mdet branches can be reached? Is the det branch reached first?
How about removing Deterministic for good and properly adapting this logic?

File: framework/tests/flow2/Purity.java [hunk 6]
```diff
@@ -26,6 +28,19 @@ public PureClass() {}
...
+    @MultipleRunDeterministic
+    void mdet() {}
+
+    @SingleRunDeterministic
+    void sdet() {}
+
+    @SingleRunDeterministic
+    void bad() {
+        //:: error: (purity.not.singlerundeterminisitic.call)
+        mdet();
```
Comment (wmdietl) [inline framework/tests/flow2/Purity.java]: Isn't SingleRunDeterministic weaker than MultiRunDeterministic? Shouldn't it be safe to call a MRD method from within a SRD method?

File: dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java (hunk 5 above)
Comment (andychlee) [inline dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java]: I did try setting branches completely separate from the already implemented deterministic annotation, and I still had the same problem. I will update you once I completely remove the already existing annotation and replace all the current tests as well.

File: checker/examples/units-extension/qual/FrequencyRelations.java (hunk 3 above)
Comment (andychlee) [inline checker/examples/units-extension/qual/FrequencyRelations.java]: I think ant reformat changed the format for some reason.
I'll look into why that happened and let you know.

File: dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java (hunk 4 above)
Comment (andychlee) [inline dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java]: Yep, I'll remove all the pre-existing methods and fields.

File: framework/tests/flow2/Purity.java [hunk 7]
```diff
@@ -26,6 +28,19 @@ public PureClass() {}
         public SEClass() {}
     }
 
+    // sample test checking for deterministic
+    @MultipleRunDeterministic
```
Comment (wmdietl) [inline framework/tests/flow2/Purity.java]: There are very few tests. Did you maybe not add additional test files? In particular tests that highlight the use of MRD for loops are important.

**Comment by wmdietl** [comment]
Hi @andychlee , you should be able to use `AnnotatedTypeFactory.addAliasedDeclAnnotation` to declare an alias from `@Deterministic` to the new longer name. Then you don't need to adapt all test cases and just add new ones that use SRD.

**Comment by wmdietl** [comment]
Thanks for your updates. Can you make this as a PR against typetools/master, to make it easier to keep up-to-date?

**Comment by wmdietl** [comment]
I've updated opprop to be in sync with typetools. This PR still shows many unrelated changes. It might be best to close this PR and start afresh, either here or on typetools.

**Comment by andychlee** [comment]
I'll make a new fork and start afresh like you mentioned. I've also had Travis failing builds which may be due to sync issues with typetools.


-Andy

________________________________
